README.md
what did i load?
---

Puppetter based resource tracker. Generates a report of image/script usage across different devices

# CLI 
`npx what-did-i-load http://mysite.com`

to only collect informatio about images, but ignore some "noisy" hosts
`npx what-did-i-load http://mysite.com -f image -w 1280 -pt false -dpr 2 --ignore-hosts www.facebook.com www.google-analytics.com www.google.com
`

```
npx what-did-i-load --help

Options:
  -w, --width <number>                 width (default: "1024")
  -h, --height <number>                height (default: "768")
  -dpr, --device-pixel-ratio <number>  device pixel ratio (default: "1")
  -ph, --per-host <flag>               list information per host (default: true)
  -pt, --per-type <flag>               list information per resource type (default: true)
  -f, --focus <type>                   focus on specific type
  --host <string>                      focus on specific host
  -lh, --ignore-hosts <hosts...>       ignore set of host
  --help                               display help for command
```


# API
`import {measure, createLocalBrowser} from 'what-did-i-load'`;

- `createLocalBrowser(width, height, dpr)` - creates a puppeteer instance with a given viewport
- `measure(browser, url)` - returns all resources requested by a given page

# AWS
In order to run in AWS/Vercel environment use createAWSBrowser command. 

# See also
- https://vercel.com/thekashey/flow-meter

# License
MIT


CHANGELOG.md
## 0.0.1 (2021-05-15)





src/measure.ts
import type{FactoredBrowser} from "./createBrowser";
// import {resolve} from 'path';

type Resource = {
  url: string;
  type: string;
  size: number;
  host: string
}

export async function measure(browser: FactoredBrowser, targetUrl: string): Promise<Resource[]> {
  const page = await browser.newPage();
  await page.setCacheEnabled(false);
  const resources: Resource[] = [];

  page.on('response', async (response) => {
    const url = response.url();
    const host = new URL(url).hostname;

    try {
      resources.push({
        host,
        url,
        type: response.headers()['content-type'],
        size: (await response.buffer()).length,
      })
    } catch (e) {
      // suppress
      // console.error('error processing', url);
    }
  });

  await page.goto(targetUrl, {waitUntil: 'networkidle2'});

  await page.close();
  browser.close();

  return resources;
}

src/createBrowser.ts
import chrome from 'chrome-aws-lambda';
import puppeteerCore from 'puppeteer-core';
import puppeteer from 'puppeteer';

export type FactoredBrowser= puppeteer.Browser;
export type BrowserFactory = (viewportWidth: number, viewportHeight: number, drp: number) => Promise<FactoredBrowser>;


export const createLocalBrowser: BrowserFactory = async (viewportWidth, viewportHeight, drp) => (
  puppeteer.launch({
    headless: true,
    defaultViewport: {
      width: viewportWidth,
      height: viewportHeight,
      deviceScaleFactor: drp
    }
  })
)

export const createAWSBrowser: BrowserFactory = async (viewportWidth, viewportHeight, dpr) => (
  puppeteerCore.launch({
    args: chrome.args,
    executablePath: await chrome.executablePath,
    headless: true,
    defaultViewport: {
      width: viewportWidth,
      height: viewportHeight,
      deviceScaleFactor: dpr
    }
  }) as any
)

src/cli.ts
#!/usr/bin/env node

import {Command} from 'commander';
import {measure} from "./index";
import {createLocalBrowser} from "./createBrowser";

const program = new Command();

const getType = (mime: string = 'undefined'): string => {
  if (mime.startsWith('image/svg')) {
    return 'svg';
  }
  if (mime.startsWith('image/')) {
    return 'image';
  }
  if (mime.startsWith('application/javascript') || mime.startsWith('application/x-javascript') || mime.startsWith('text/javascript')) {
    return 'script';
  }
  if (mime.startsWith('text/css')) {
    return 'style';
  }
  return mime.split(';')[0];
}

const parseBool = (x: string): boolean => x !== 'false'

program
  .arguments('<url>')
  .option('-w, --width <number>', 'width', '1024')
  .option('-h, --height <number>', 'height', '768')
  .option('-dpr, --device-pixel-ratio <number>', 'device pixel ratio', '1')
  .option('-ph, --per-host <flag>', 'list information per host', parseBool, true)
  .option('-pt, --per-type <flag>', 'list information per resource type', parseBool, true)
  .option('-f, --focus <type>', 'focus on specific type')
  .option('--host <string>', 'focus on specific host')
  .option('-lh, --ignore-hosts <hosts...>', 'ignore set of host')
  .action(async (url, options) => {
    measure(
      await createLocalBrowser(+options.width, +options.height, +options.devicePixelRatio),
      url
    ).then((data) => {
      const types: Record<string, Record<string, number>> = {};
      const perType: Record<string, { count: number, bytes: number }> = {};

      const typeRec = new Set<string>();
      data.forEach((resource) => {
        typeRec.add(getType(resource.type));
      });
      // using presorted types to keep result always ordered in the same way
      Array.from(typeRec.values()).sort().forEach(type => {
        types[type] = {};
        perType[type] = {
          count: 0,
          bytes: 0,
        };
      })

      data.forEach((resource) => {
        const type = getType(resource.type);
        if (options.focus && options.focus !== type) {
          return;
        }
        if (options.ignoreHosts && options.ignoreHosts.includes(resource.host)) {
          return;
        }
        if (options.host && options.host !== resource.host) {
          return;
        }

        if (!types[type][resource.host]) {
          types[type][resource.host] = 0;
        }
        types[type][resource.host] += resource.size;
        perType[type].bytes += resource.size;
        perType[type].count += 1;
      });
      if (options.perType) {
        console.log(JSON.stringify(perType, undefined, 2))
      }
      if (options.perHost) {
        console.log(JSON.stringify(types, undefined, 2))
      }
    })
      .then(() => process.exit(0))
  });

program.addHelpText('after', `

Example call:
  $ what-did-i-load https://facebook.com`);

program.parse();